File: app/src/dev/java/com/github/abhijitpparate/keeps/data/database/FirebaseDatabaseService.java
package com.github.abhijitpparate.keeps.data.database;

import android.support.annotation.NonNull;
import android.util.Log;

import com.github.abhijitpparate.keeps.data.auth.User;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.CompletableEmitter;
import io.reactivex.CompletableOnSubscribe;
import io.reactivex.Maybe;
import io.reactivex.MaybeEmitter;
import io.reactivex.MaybeOnSubscribe;

public class FirebaseDatabaseService implements DatabaseSource {

    private static final String USERS = "USERS";
    private static final String NOTES = "NOTES";

    private FirebaseDatabaseService() {

    }

    public static DatabaseSource getInstance() {
        return new FirebaseDatabaseService();
    }

    @Override
    public Completable createProfile(final Profile profile) {
        return Completable.create(
                new CompletableOnSubscribe() {
                    @Override
                    public void subscribe(final CompletableEmitter e) throws Exception {
                        final DatabaseReference rootRef = FirebaseDatabase.getInstance().getReference();
                        final DatabaseReference idRef = rootRef.child(USERS).child(profile.getUid());
                        idRef.addListenerForSingleValueEvent(new ValueEventListener() {
                            @Override
                            public void onDataChange(DataSnapshot snapshot) {
                                if (!snapshot.exists()) {
                                    idRef.setValue(profile)
                                    .addOnCompleteListener(new OnCompleteListener<Void>() {
                                        @Override
                                        public void onComplete(@NonNull Task<Void> task) {
                                            if (task.isSuccessful()) {
                                                e.onComplete();
                                            } else {
                                                e.onError(task.getException());
                                            }
                                        }
                                    });
                                } else {
                                    e.onComplete();
                                }
                            }

                            @Override
                            public void onCancelled(DatabaseError databaseError) {
                                Log.d("FIREBASE", databaseError.toString());
                            }
                        });
                    }
                }
        );
    }

    @Override
    public Maybe<Profile> getProfile(final String uid) {
        return Maybe.create(
                new MaybeOnSubscribe<Profile>() {
                    @Override
                    public void subscribe(final MaybeEmitter<Profile> e) throws Exception {
                        final DatabaseReference rootRef = FirebaseDatabase.getInstance().getReference();
                        DatabaseReference idRef = rootRef.child(USERS).child(uid);
                        idRef.addListenerForSingleValueEvent(new ValueEventListener() {
                            //does this check node for activeUser exists?
                            @Override
                            public void onDataChange(DataSnapshot snapshot) {
                                if (snapshot.exists()) {
                                    Profile profile = snapshot.getValue(Profile.class);
                                    e.onSuccess(profile);
                                } else {
                                    e.onComplete();
                                }
                            }

                            @Override
                            public void onCancelled(DatabaseError databaseError) {
                                Log.d("FIREBASE", databaseError.toString());
                            }
                        });
                    }
                }
        );
    }

    @Override
    public Maybe<Note> getNoteFromId(final String uid, final String noteId) {
        return Maybe.create(
                new MaybeOnSubscribe<Note>() {
                    @Override
                    public void subscribe(final MaybeEmitter<Note> e) throws Exception {
                        final DatabaseReference rootRef = FirebaseDatabase.getInstance().getReference();
                        DatabaseReference idRef = rootRef.child(USERS).child(uid).child(NOTES).child(noteId);
                        idRef.addListenerForSingleValueEvent(new ValueEventListener() {
                            //does this check node for activeUser exists?
                            @Override
                            public void onDataChange(DataSnapshot snapshot) {
                                if (snapshot.exists()) {
                                    Note note = snapshot.getValue(Note.class);
                                    e.onSuccess(note);
                                } else {
                                    e.onComplete();
                                }
                            }

                            @Override
                            public void onCancelled(DatabaseError databaseError) {
                                Log.d("FIREBASE", databaseError.toString());
                            }
                        });
                    }
                }
        );
    }

    @Override
    public Maybe<List<Note>> getNotesForCurrentUser(final String uid) {
        return Maybe.create(
                new MaybeOnSubscribe<List<Note>>() {
                    @Override
                    public void subscribe(final MaybeEmitter<List<Note>> e) throws Exception {
                        final DatabaseReference rootRef = FirebaseDatabase.getInstance().getReference();
                        DatabaseReference idRef = rootRef.child(USERS).child(uid).child(NOTES);
                        idRef.addListenerForSingleValueEvent(new ValueEventListener() {
                            //does this check node for activeUser exists?
                            @Override
                            public void onDataChange(DataSnapshot snapshot) {
                                if (snapshot.exists()) {
                                    List<Note> notes = new ArrayList<>();
                                    for (DataSnapshot postSnapshot: snapshot.getChildren()){
                                        Log.d("FIREBASE", "onDataChange: " + postSnapshot.toString());
                                        Note n = postSnapshot.getValue(Note.class);
                                        notes.add(n);
                                    }
                                    e.onSuccess(notes);
                                } else {
                                    e.onComplete();
                                }
                            }

                            @Override
                            public void onCancelled(DatabaseError databaseError) {
                                Log.d("FIREBASE", databaseError.toString());
                            }
                        });
                    }
                }
        );
    }

    @Override
    public Completable createNewNote(final String uid, final Note note) {
        return Completable.create(
                new CompletableOnSubscribe() {
                    @Override
                    public void subscribe(@io.reactivex.annotations.NonNull final CompletableEmitter e) throws Exception {
                        final DatabaseReference rootRef = FirebaseDatabase.getInstance().getReference();
                        final DatabaseReference idRef = rootRef.child(USERS).child(uid).child(NOTES).child(note.getNoteId());
                        idRef.addListenerForSingleValueEvent(new ValueEventListener() {
                            @Override
                            public void onDataChange(DataSnapshot snapshot) {
                                if (!snapshot.exists()) {
                                    idRef.setValue(note)
                                            .addOnCompleteListener(new OnCompleteListener<Void>() {
                                                @Override
                                                public void onComplete(@NonNull Task<Void> task) {
                                                    if (task.isSuccessful()) {
                                                        e.onComplete();
                                                    } else {
                                                        e.onError(task.getException());
                                                    }
                                                }
                                            });
                                } else {
                                    e.onComplete();
                                }
                            }

                            @Override
                            public void onCancelled(DatabaseError databaseError) {
                                Log.d("FIREBASE", databaseError.toString());
                            }
                        });
                    }
                }
        );
    }

    @Override
    public Completable deleteProfile(final String uid) {
        return Completable.create(
                new CompletableOnSubscribe() {
                    @Override
                    public void subscribe(final CompletableEmitter e) throws Exception {
                        final DatabaseReference rootRef = FirebaseDatabase.getInstance().getReference();
                        rootRef.child(USERS)
                                .child(uid)
                                .setValue(null)
                                .addOnCompleteListener(new OnCompleteListener<Void>() {
                                    @Override
                                    public void onComplete(@NonNull Task<Void> task) {
                                        if (task.isSuccessful()) {
                                            e.onComplete();
                                        } else {
                                            e.onError(task.getException());
                                        }
                                    }
                                });
                    }

                });
    }

    @Override
    public Completable updateProfile(final Profile profile) {
        return Completable.create(
                new CompletableOnSubscribe() {
                    @Override
                    public void subscribe(final CompletableEmitter e) throws Exception {
                        final DatabaseReference rootRef = FirebaseDatabase.getInstance().getReference();
                        rootRef.child(USERS)
                                .child(profile.getUid())
                                .setValue(profile)
                                .addOnCompleteListener(new OnCompleteListener<Void>() {
                                    @Override
                                    public void onComplete(@NonNull Task<Void> task) {
                                        if (task.isSuccessful()) {
                                            e.onComplete();
                                        } else {
                                            e.onError(task.getException());
                                        }
                                    }
                                });
                    }

                });
    }

    @Override
    public void setReturnFail(boolean bool) {
        //only for testing purposes
    }

    @Override
    public void setReturnEmpty(boolean bool) {
        //only for testing purposes

    }

}

File: app/src/dev/java/com/github/abhijitpparate/keeps/data/database/FirebaseDatabaseService.kt
package com.github.abhijitpparate.keeps.data.database

import android.util.Log
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import io.reactivex.Completable
import io.reactivex.Maybe
import java.lang.Error
import java.util.*

class FirebaseDatabaseService private constructor() : DatabaseSource {

    override fun createProfile(profile: Profile): Completable {
        return Completable.create { e ->
            val rootRef = FirebaseDatabase.getInstance().reference
            val idRef = rootRef.child(USERS).child(profile.uid)
            idRef.addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (!snapshot.exists()) {
                        idRef.setValue(profile)
                                .addOnCompleteListener { task ->
                                    if (task.isSuccessful) {
                                        e.onComplete()
                                    } else {
                                        e.onError(task.exception)
                                    }
                                }
                    } else {
                        e.onComplete()
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    Log.d("FIREBASE", databaseError.toString())
                    e.onError(Error("Database error"))
                }
            })
        }
    }

    override fun getProfile(uid: String): Maybe<Profile> {
        return Maybe.create { e ->
            val rootRef = FirebaseDatabase.getInstance().reference
            val idRef = rootRef.child(USERS).child(uid)
            idRef.addListenerForSingleValueEvent(object : ValueEventListener {
                //does this check node for activeUser exists?
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (snapshot.exists()) {
                        val profile = snapshot.getValue(Profile::class.java)
                        e.onSuccess(profile)
                    } else {
                        e.onError(Error("Profile not found"))
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    Log.d("FIREBASE", databaseError.toString())
                    e.onError(Error("Database error"))
                }
            })
        }
    }

    override fun getNoteFromId(uid: String, noteId: String): Maybe<Note> {
        return Maybe.create { e ->
            val rootRef = FirebaseDatabase.getInstance().reference
            val idRef = rootRef.child(USERS).child(uid).child(NOTES).child(noteId)
            idRef.addListenerForSingleValueEvent(object : ValueEventListener {
                //does this check node for activeUser exists?
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (snapshot.exists()) {
                        val note = snapshot.getValue(Note::class.java)
                        e.onSuccess(note)
                    } else {
                        e.onError(Error("Note not found"))
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    Log.d("FIREBASE", databaseError.toString())
                    e.onError(Error("Database error"))
                }
            })
        }
    }

    override fun getNotesForCurrentUser(uid: String): Maybe<List<Note>> {
        return Maybe.create { e ->
            val rootRef = FirebaseDatabase.getInstance().reference
            val idRef = rootRef.child(USERS).child(uid).child(NOTES)
            idRef.addListenerForSingleValueEvent(object : ValueEventListener {
                //does this check node for activeUser exists?
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (snapshot.exists()) {
                        val notes = ArrayList<Note>()
                        for (postSnapshot in snapshot.children) {
                            Log.d("FIREBASE", "onDataChange: " + postSnapshot.toString())
                            val n = postSnapshot.getValue(Note::class.java) as Note
                            notes.add(n)
                        }
                        e.onSuccess(notes)
                    } else {
                        e.onError(Error("Notes not found"))
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    Log.d("FIREBASE", databaseError.toString())
                }
            })
        }
    }

    override fun createOrUpdateNote(uid: String, note: Note): Completable {
        return Completable.create { e ->
            val rootRef = FirebaseDatabase.getInstance().reference
            val idRef = rootRef.child(USERS).child(uid).child(NOTES).child(note.noteId)
            idRef.addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    idRef.setValue(note)
                            .addOnCompleteListener { task ->
                                if (task.isSuccessful) {
                                    e.onComplete()
                                } else {
                                    e.onError(task.exception)
                                }
                            }
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    Log.d("FIREBASE", databaseError.toString())
                }
            })
        }
    }

    override fun deleteProfile(uid: String): Completable {
        return Completable.create { e ->
            val rootRef = FirebaseDatabase.getInstance().reference
            rootRef.child(USERS)
                    .child(uid)
                    .setValue(null)
                    .addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            e.onComplete()
                        } else {
                            e.onError(task.exception)
                        }
                    }
        }
    }

    override fun updateProfile(profile: Profile): Completable {
        return Completable.create { e ->
            val rootRef = FirebaseDatabase.getInstance().reference
            rootRef.child(USERS)
                    .child(profile.uid)
                    .setValue(profile)
                    .addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            e.onComplete()
                        } else {
                            e.onError(task.exception)
                        }
                    }
        }
    }

    override fun setReturnFail(bool: Boolean) {
        //only for testing purposes
    }

    override fun setReturnEmpty(bool: Boolean) {
        //only for testing purposes

    }

    companion object {

        private val USERS = "USERS"
        private val NOTES = "NOTES"

        val instance: DatabaseSource
            get() = FirebaseDatabaseService()
    }

}